tests: Add random action injector

abhijat opened the pull request:
## Cover letter

Adds random action injector utilities, aimed to be used for e2e tests. The failures could be process failures, node decommission, leadership transfer etc, controlled by a context manager.

The action injector runs on a thread and periodically introduces changes on randomly selected nodes in the cluster.

### Features

* The new context can be used as following

```
with random_process_kills(self.redpanda):
    do_something()
```

Changes in [force push](https://github.com/redpanda-data/redpanda/compare/7a0ccb71ed770c8588fb318bc72f2a6f16af0aab..d5f909d73a346e55d37e8a425d6e535721c2515d)

* clean up unused code based on review comments
* simplify test cases based on review comments

Changes in [force push](https://github.com/redpanda-data/redpanda/compare/d5f909d73a346e55d37e8a425d6e535721c2515d..66afa5bce0955f6f09f64bb1305394ea7aa25f53)
* clean up result handling

Changes in [force push](https://github.com/redpanda-data/redpanda/compare/66afa5bce0955f6f09f64bb1305394ea7aa25f53..28cfb562f1aa98ba07ea83c8d7767f74d7a44d2c):
* resolve conflicts with upstream dev

Changes in [force push](https://github.com/redpanda-data/redpanda/compare/035a92102eb5ea1ca16469c3e29c7580d44c9575..40b3d0fdbc21d0d46e18d2a1687e81fdafb2780d)
* fold everything back into two logical commits
* remove changes which were only import reordering

Changes in [force push](https://github.com/redpanda-data/redpanda/compare/3d95f859f306665151f12e38fa49e6741201b21c..3d6686f770549a3896687489f376f8168482df34)
* system restoration code removed
* fix assertion to only check if action was triggered by test

jcsp commented on tests/rptest/tests/e2e_shadow_indexing_test.py:
```diff
@@ -107,6 +117,41 @@ def test_write(self):
+    override_default_topic_replication = EndToEndShadowIndexingBase.num_brokers
+
+    @cluster(num_nodes=5,
+             log_allow_list=RESTART_LOG_ALLOW_LIST,
```
You'll probably find you need CHAOS_LOG_ALLOW_LIST here (although awesome if it works without it)

abhijat replied:
I just ran into a could not parse header - will update the list

graphcareful reviewed (dismissed):
Really nice job, LGTM

jcsp reviewed (commented):
Structure looks great.

Some comments - I think I might have commented on code that isn't used yet (e.g. the decom/recom actions)

jcsp commented:
Thank you for adding release notes, but those are usually used for customer-facing release documentation, so we generally don't mention changes to our internal test code.

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+            self,
+            config: ActionConfig,
+            redpanda: RedpandaService,
+            random_op: RandomNodeOp,
```
`random_op` is a misleading name

abhijat replied:
I have changed the name to `disruptive_action`, let me know your thoughts on this.

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+        self.redpanda = redpanda
+        self.nodes_affected = set()
+
+    def limit_reached(self) -> bool:
```
`limit_reached` is a very genetic name it's unclear which limit you refer to. maybe it makes sense to add a comment and/or rename it

abhijat replied:
Thanks for the suggestion, I have renamed it to `max_affected_nodes_reached` and also added a docstring

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+
+class RandomNodeDecommission(RandomNodeOp):
+    def limit_reached(self):
+        return len(self.nodes_affected) >= self.config.max_affected_nodes
```
we have `nodes_affected` and `max_affected_nodes` let's rename the former to affected_nodes

abhijat replied:
fixed

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+    max_time_between_actions_sec: float
+    max_affected_nodes: Optional[int] = None
+
+    def random_time_in_range(self) -> float:
```
Let's rename `random_time_in_range` to `time_between_actions`, it gives more context about method and explains what it's for

abhijat replied:
fixed

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+    def action(self):
+        raise NotImplementedError
+
+    def __call__(self, *args, **kwargs):
```
I have removed the `__call__` override and replaced it with direct calls to `action` and `reverse_action`, the checks are done in these methods now and there are `do_..` counterparts which child classes can override.

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+        return random.uniform(self.min_time_between_actions_sec, self.max_time_between_actions_sec)
+
+
+class RandomNodeOp:
```
Thanks for the suggestion, I see the name is a bit disambiguous. I decided to rename it to `DisruptiveAction` which I think is closer to the intent but suggestions are welcome. I think `PeriodicAction` could be used as well but it does not convey _what_ the action does. 

I did not want to name it something related to Failure because the action could be something which is not failure but simply adding a topic or deleting a topic, or leadership transfer etc.

> If you move node selection to ActionInjectorThread and just pass a node to the action method you'll reduce cognitive load: less methods => less possible interactions.

I had node selection in the thread in an early version of the code, but I moved it to the action because in some cases node may not be required for the action being performed eg creating a topic. So we let the action decide what it wants to do for node selection. 

There are fields in `DisruptiveAction` which are node specific but it is up to the action to use them or not.

rystsov replied:
I've looked through the changes and in all the cases we use the base class's target so I can't find the case you mention.

abhijat replied:
It was the case for leadership transfer [here](https://github.com/redpanda-data/redpanda/commit/ce5caaeb0b749242ccf9b8b3634f64d813b1927e#diff-d0b45de9466a84e1f2fb8c4de043a364b399cc796c2653b84922ad353c49a3feR69) which I removed from this PR but I am working on in another PR.

Node will also not be required for topic create and delete action which I have to work on next. Neither will require node to be passed to action.

rystsov replied:
I still see `RandomNodeOp` in the first commit instead of `DisruptiveAction`; you rename RandomNodeOp to DisruptiveAction only in 4th commit "tests: attempt to restore state on exit". Please edit the history and do not introduce `RandomNodeOp` in the first place. This comment also applies to `__call__`, `limit_reached` and other similar issues

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -111,73 +121,104 @@ def limit_reached(self):
+            # Update started_nodes so validations are run on the correct
+            # set of nodes later.
+            try:
+                self.redpanda.started_nodes().remove(node)
```
I have added a couple of methods to the `Redpanda` class to modify the state explicitly.

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -160,7 +213,15 @@ def run(self):
         while not self._stop_requested.is_set():
-            self.random_op()
+            if self.config.reverse_action_on_next_cycle:
+                result = self.random_op(reverse=True)
```
I have changed the code so that the methods are now called directly and removed the flag.

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+        self.redpanda = redpanda
+        self.nodes_affected = set()
+
+    def limit_reached(self) -> bool:
```
Thanks, I have added docstrings to most of the methods and also renamed some methods to make things a little clearer.

abhijat commented on tests/rptest/services/redpanda.py:
```diff
@@ -911,7 +911,7 @@ def raise_on_crash(self):
                 crashes.append((node, line))
 
-        if not crashes:
+        if not crashes and not allow_missing_process:
```
This makes sense, I am working on adding changes so that system state can optionally be restored on context exit. The actions will need to decide how they want to restore the changes made. I will add the changes to this same PR and remove this new flag.

abhijat replied:
I have added changes to restore the state now, and also removed the `allow_missing_process` flag

abhijat commented on tests/rptest/tests/e2e_shadow_indexing_test.py:
```diff
@@ -44,23 +49,29 @@ def __init__(self, test_context):
 
         si_settings.load_context(self.logger, test_context)
 
+        if self.override_default_topic_replication:
```
When I was running this test on repeat, I saw the test fail a few times when waiting on consumer. Looking at the consumer log I saw the following errors repeated (on failing test instances):

```
[2022-04-27 06:30:11,486] WARN [Consumer clientId=consumer-test_group-1, groupId=test_group] Connection to node 2147483646 (docker_n_1/172.18.0.5:9092) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient)
[2022-04-27 06:30:11,486] INFO [Consumer clientId=consumer-test_group-1, groupId=test_group] Group coordinator docker_n_1:9092 (id: 2147483646 rack: null) is unavailable or invalid due to cause: null.isDisconnected: true. Rediscovery will be attempted. (org.apache.kafka.clients.consumer.internals.AbstractCoordinator)
[2022-04-27 06:30:11,486] INFO [Consumer clientId=consumer-test_group-1, groupId=test_group] Rebalance failed. (org.apache.kafka.clients.consumer.internals.AbstractCoordinator)
```

It seems on researching that error, that when group coordinator is killed the tests fail because `__consumer_offsets` is not replicated by default and there is no resilience (please correct me if I am mistaken), so I set replication for all topics by default to number of brokers, after which the tests pass correctly even if the coordinator is killed.

I could not find a specific setting to control the replication for the consumer offsets topic.

I have cleaned up the code so that the replication count is set up in the specific test where it is required and not spread across the base and child classes.

abhijat commented:
@jcsp @rystsov I will focus this PR on the action which kills processes. I have reduced the NodeDecommission and LeadershipTransfer actions to skeleton classes, and will focus on concrete implementation for them in subsequent PRs after more research based on the comments here. 

If it is better that I should remove the skeleton classes altogether, please let me know.

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,183 @@
+    def action(self):
+        node = self.target_node()
+        if node:
+            broker_id = next((i for i, n in enumerate(self.redpanda.nodes) if n == node))
```
Thanks for the info, I have changed the node decommission and leadership transfer to skeleton classes for now, it looks like more research is needed on my part before implementing them. 

I will put in concrete implementations for these and possibly other actions in future PRs. The rptest in this PR only uses process killing with the latest commit.

If it is better to remove these actions for now let me know.

graphcareful commented:
One comment about our PR flow, when iterating we don't push commits on top of already reviewed commits, we just edit/rebase existing commits. If you rebase with `--keep-base` reviewers will be able to see the diff of the PR between force pushes.

jcsp reviewed (dismissed):
LGTM when CI passes: I like the overall structure + am okay with having the not-yet-implemented actions in there as stubs.

Needs a re-check from @rystsov

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,330 @@
+        last_affected_node, self.last_affected_node = self.last_affected_node, None
+        return last_affected_node
+
+    def do_restore_nodes(self, nodes_to_restore: Set[ClusterNode]):
+        """
+        Attempt to restore the redpanda process on all nodes where it was stopped.
+        """
+        for node in nodes_to_restore:
+            self._start_rp(node)
```
this looks error prone, during each action we return an affected node, all the affected nodes are accumulated to action_log which eventually is passed to do_restore_nodes; the problem is that ProcessKill is reversible and all but the last actions are already reverted; so we'll try to reverse it twice;

I recommend to remove the do_restore_nodes, restore_state_on_exit, action_log and all the corresponding config and use just the restore method

abhijat replied:
The `restore_nodes` function makes sure that only the nodes which have not been effectively reversed are restored, please see comment above. The log is processed in `restore_nodes` to find which nodes have already been recovered and skips them.

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,330 @@
+        self.redpanda.logger.info('cluster is ready, starting action loop')
+
+        while not self._stop_requested.is_set():
+            if self.config.reverse_action_on_next_cycle:
```
this looks error prone, we already have is_reversible property on the DisruptiveAction, as I understand reverse_action_on_next_cycle always should have the same value as is_reversible, so when we forget about this link it creates an issue

I suggest to remove reverse_action_on_next_cycle and rely on is_reversible

abhijat replied:
thanks for the suggestion, I have cleaned up the extra flag from config and now we rely solely on is_reversible

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,330 @@
+
+        while not self._stop_requested.is_set():
+            if self.config.reverse_action_on_next_cycle:
+                result = self.disruptive_action.reverse()
```
this is another questionable piece, it seems that we always invoke reverse before invoking action it means that for the first iteration we ask an action to reverse while it wasn't applied yet. I'd guessed that we ask a system to undo what it hasn't done yet it should throw an error but we need to tolerate it

I suggest to use something like:

```
while ...:
  disruptive_action.action()
  sleep(time_between_actions())
  if disruptive_action.is_reversible:
    disruptive_action.reverse()
```

abhijat replied:
fixed

rystsov commented on tests/rptest/services/admin.py:
```diff
@@ -419,6 +419,14 @@ def decommission_broker(self, id, node=None):
         self.redpanda.logger.debug(f"decommissioning {path}")
         return self._request('put', path, node=node)
 
+    def recommission_broker(self, broker_id, node=None):
```
isn't used, please remove

rystsov commented on tests/rptest/tests/e2e_shadow_indexing_test.py:
```diff
@@ -85,6 +96,56 @@ def test_write(self):
+
+
+class EndToEndShadowIndexingTestWithDisruptions(EndToEndShadowIndexingBase):
+    def _build_redpanda_instance(self):
```
Polymorphism is a useful thing but it has a cost (jumping between classes to understand the context) so when we may get away without it - it's better to avoid it. In this case we need _build_redpanda_instance only to pass extra config via extra_rp_conf. You can do the same just by passing extra_rp_conf via constructor like we do in the other tests inherited from RedpandaTest.

abhijat replied:
thanks for the suggestion, I have made changes as suggested, extra_rp_conf now passed via constructor

rystsov reviewed (changes requested):
It's better to remove the code we don't use. Code has gravity so the more we push the harder it change later; also it creates more load on the reviewer and if we never follow up it's instant dead code.

Also please follow @graphcareful advice and rewrite the history, we tend to care about the shape of the commit history - https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#commit-history

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,330 @@
+        Optionally restore state when the action injector thread is ending.
+        Uses the action log to determine what restoration should be done.
+        """
+        node_final_state = defaultdict(lambda: False)
+        for entry in action_log:
+            node_final_state[entry.node] = entry.is_reverse_action
+
+        nodes_to_restore = {
+            node
+            for node, is_reversed in node_final_state.items()
+            if not is_reversed
+        }
```
The map is to capture the final state of each node and to avoid restore on nodes which do not need it. If action log looks like this

```
[
 (NodeA, Down),
 (NodeA, Up),
 (NodeB, Down),
 (NodeB, Up),
 (NodeC, Down),
]
```

then the map will look like `{NodeA: Up, NodeB: Up, NodeC: Down}` and the set will contain just NodeC, which is the only node to bring up.

It could also be possible to just look at the last entry in action log, but it is possible that some recovery has failed at some point and multiple nodes are down, so it is safer to look at the entire list, calculate which actions were finally not reversed, and reverse them.

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,330 @@
+        self.do_restore_nodes(nodes_to_restore)
+
+
+class NodeDecommission(DisruptiveAction):
```
thanks, removed this and other pieces of unused code. will create a separate PR later

abhijat commented:
> One comment about our PR flow, when iterating we don't push commits on top of already reviewed commits, we just edit/rebase existing commits. If you rebase with `--keep-base` reviewers will be able to see the diff of the PR between force pushes.

@graphcareful I have used force push with `--keep-base` for the latest commit editing the last approved commit, please let me know if the history looks more acceptable now. 

I have also added links to description of force push in the cover letter after discussing with Evgeny so it is easier for reviewers.

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,279 @@
+                self.action_log.append(
+                    ActionLogEntry(result, is_reverse_action=True))
+
+        if self.config.restore_state_on_exit:
+            self.redpanda.logger.info('attempting to restore system state')
+            self.disruptive_action.restore_state_on_exit(self.action_log)
```
It seems that this code doesn't do anything: we already invoke `.reverse` for each `.action` so by the time we reached this method all the actions are already reversed.

As a result we don't need:
 - ActionConfig::restore_state_on_exit (config)
 - DisruptiveAction::restore_state_on_exit (method)
 - `action_log` (tracking history of a DisruptiveAction's actions inside ActionInjectorThread)
 - return nodes from `.reverse` and `.action`
 - introduce ActionLogEntry

It's better avoid code which we don't need

abhijat replied:
Fair point. When I added the restore action I was thinking of operations which are not reversible but still need some kind of restore on exit. But now that you pointed it out I realised that there are no such operations that will require such a restore at this time that I can think of.

If there are such operations in future, I will revisit, but I will remove this code for now. 

Action log might still be required though because it allows assertions by unit test code easily to make sure that the code is called. 

I will check if some intercept can be used to assert that the action injector has worked, otherwise I might need to keep the log for assertions (or some simpler mechanism than a log).

abhijat replied:
removed the restore related code in latest commit. for making assertions easy I have changed it to two flags and an assert helper on the action context, and the log is removed

rystsov commented on tests/rptest/tests/e2e_shadow_indexing_test.py:
```diff
@@ -95,11 +99,56 @@ def test_write(self):
+            self.start_consumer()
+            self.run_validation()
+
+        action_log = ctx.action_log()
+        assert action_log
+
+        # At least one process kill action is present
+        assert action_log[0].is_reverse_action is False
+
+        # If there was another action after process kill, it was revival,
+        # because we are running in reverse after action mode
+        if len(action_log) > 1:
+            assert action_log[1].is_reverse_action and action_log[
+                1].node == action_log[0].node
```
this check is excessive especially if we remove `restore_state_on_exit` and it requires to track action_log as part of `ActionInjectorThread` we may achieve the same effect if `ProcessKill` internally asserts that `.reverse` must follow `.action`

abhijat replied:
Sure, will cover this as part of the general code change with removing restore etc.

abhijat replied:
fixed in latest commit

rystsov reviewed (commented):
Thanks @abhijat it's way easier to read after you edited the history. PR looks good, the only suspension place is `restore_state_on_exit`, it seems we don't need it and we may avoid introducing several concepts which were added just to support it

abhijat commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,279 @@
+            self.redpanda.logger.info(
+                f'selected {selected.account.hostname} of {names} for operation'
+            )
+            return selected
```
fixed in latest commit and updated the method type signature

abhijat commented:
@rystsov please review, I have cleaned up the restoration code and replaced log with boolean that the test can use to assert internal state of the thread.

rystsov commented on tests/rptest/services/action_injector.py:
```diff
@@ -0,0 +1,238 @@
+
+class ActionCtx:
+    def __init__(self, config: ActionConfig, redpanda: RedpandaService,
+                 random_op: DisruptiveAction):
```
nit: random_op -> disruptive_action

abhijat replied:
Fixed in latest commit

rystsov reviewed (approved):
Looks good!

rystsov commented:
Some of the tests are failing, please for each failing test check if it's a know [flaky test](https://github.com/redpanda-data/redpanda/issues?q=is%3Aissue+is%3Aopen++label%3Aci-failure+) if it's then add a link to this build to the flaky issue and comment this PR to reassure that the failing tests aren't related to this PR.

For new failing tests you should investigate if it's related to this PR. If they are - fix the PR, if they aren't then you should open ci-failure tagged issues.

abhijat commented:
new issues #4601 and #4602 for ci failures seen

jcsp reviewed (approved):
This is good to go from my pov

abhijat commented:
/backport v22.1.x

vbotbuildovich commented:
Failed to run cherry-pick command. see [workflow](https://github.com/redpanda-data/redpanda/actions/runs/2285685976)
  I executed the below command:
  ```
  git cherry-pick -x 551045ec62d06c97825e804602aa2478f9d0f382 889747877f0cdc0c16ef82e9b5499a83e595c912
  ```
